### PR TITLE
Detect safe navigation operator as nil check

### PR DESCRIPTION
**Pull request by mvz**
#749.

This also improves NilCheck output when multiple lines have a nil check.

### CODE REVIEW

File: spec/reek/smells/nil_check_spec.rb [hunk 1]
```diff
@@ -19,55 +19,69 @@ def nilcheck foo
     expect('def no_nils; end').not_to reek_of(:NilCheck)
   end
 
-  it 'reports when scope uses multiple nil? methods' do
+  it 'reports when scope use #nil?' do
```
Comment (troessner) [inline spec/reek/smells/nil_check_spec.rb]: s/use/uses

File: spec/reek/smells/nil_check_spec.rb [hunk 2]
```diff
@@ -19,55 +19,69 @@ def nilcheck foo
...
       para === nil
     end
     EOS
     expect(src).to reek_of(:NilCheck)
   end
 
   it 'reports when scope uses nil ==' do
-    expect('def chk_eq_nil_rev(para); nil == para; end').to reek_of(:NilCheck)
+    src = <<-EOS
+    def chk_eq_nil_rev(para)
```
Comment (troessner) [inline spec/reek/smells/nil_check_spec.rb]: Can we rename this as well and above? The name decreases readability and hints at being relevant when it isn't.
How about just "m"?

File: spec/reek/smells/nil_check_spec.rb (hunk 1 above)
Comment (mvz) [inline spec/reek/smells/nil_check_spec.rb]: Definitely.

File: spec/reek/smells/nil_check_spec.rb [hunk 3]
```diff
@@ -19,55 +19,69 @@ def nilcheck foo
...
     src = <<-EOS
     def case_nil
       case @inst_var
       when nil then puts "Nil"
       end
-      puts "Hello"
-      case @inst_var2
-      when 1 then puts 1
-      when nil then puts nil.inspect
-      end
```
Comment (troessner) [inline spec/reek/smells/nil_check_spec.rb]: Subtle but very nice improvement.

File: spec/reek/smells/nil_check_spec.rb (hunk 2 above)
Comment (mvz) [inline spec/reek/smells/nil_check_spec.rb]: Sure.

File: spec/reek/smells/nil_check_spec.rb [hunk 4]
```diff
@@ -19,55 +19,69 @@ def nilcheck foo
...
+    def chk_multi_nil(para)
+      para&.foo
     end
     EOS
     expect(src).to reek_of(:NilCheck)
   end
+
+  it 'reports all lines when scope uses multiple nilchecks' do
+    src = <<-EOS
+    def chk_multi_nil(para)
```
Comment (troessner) [inline spec/reek/smells/nil_check_spec.rb]: Let's rename this here and above as well.

File: spec/reek/smells/nil_check_spec.rb [hunk 5]
```diff
@@ -19,55 +19,69 @@ def nilcheck foo
...
+      para.nil?
+      para === nil
```
Comment (troessner) [inline spec/reek/smells/nil_check_spec.rb]: How about adding the nil-safe operator here as well?

**Comment by troessner** [comment]
Couple of comments but looks great overall!

**Comment by troessner** [comment]
Let's also update our docs/ file.
Would also be great if our update would include some examples how to fix this smell :)

**Comment by mvz** [comment]
> Would also be great if our update would include some examples how to fix this smell :)

I'll make a new issue for that.

**Comment by mvz** [comment]
I've processed all comments, @troessner.

**Comment by troessner** [comment]
Very nice, please squash and lgtm!

**Comment by mvz** [comment]
It's actually two unrelated changes, hence the two commits.

**Comment by troessner** [comment]
Awesome! @mvz care for a release?

**Comment by troessner** [comment]
Update: Just released 4.2.5!

**Comment by mvz** [comment]
Thanks, @troessner!

**Pull request merged**